pkg/objectstore/bucket.go

```diff
@@ -145,74 +139,58 @@ func (p *provider) getOrCreateBucket(ctx context.Context, bucketName string) (Bu
 	if err == nil {
 		return d, nil
 	}
-	// Attempt creating it
+	// Attempt to create it.
 	return p.CreateBucket(ctx, bucketName)
 }
 
-type s3Provider struct {
-	*provider
+func (p *provider) bucketConfig(ctx context.Context, bucketName string) (ProviderConfig, error) {
+	if p.config.Type == ProviderTypeS3 {
+		return s3BucketConfig(ctx, p.config, p.secret, bucketName)
+	}
+	return p.config, nil
 }
 
-// Stow uses path-style requests when specifying an endpoint.
-// https://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html#path-style-access
-// https://github.com/graymeta/stow/blob/master/s3/config.go#L159
-
-const awsS3HostFmt = "https://s3.%s.amazonaws.com"
+func s3BucketConfig(ctx context.Context, c ProviderConfig, s *Secret, bucketName string) (ProviderConfig, error) {
+	if s == nil || s.Aws == nil {
+		return c, errors.New("AWS Secret required to get region")
+	}
+	r, err := s3BucketRegion(ctx, c, *s, bucketName)
+	if err != nil {
+		log.Debug().
+			WithContext(ctx).
+			WithError(err).
+			Print("Couldn't get config for bucket", field.M{"config": c})
+		return c, nil
+	}
+	c.Region = r
+	return c, nil
+}
 
-func awsS3Endpoint(region string) string {
-	return fmt.Sprintf(awsS3HostFmt, region)
+type s3Provider struct {
+	*provider
 }
 
-func (p *s3Provider) GetBucket(ctx context.Context, bucketName string) (Bucket, error) {
-	cfg := p.config
-	var err error
-	cfg.Region, err = p.getRegionForBucket(ctx, bucketName)
-	if err != nil {
-		return nil, errors.Wrapf(err, "could not get region for bucket %s", bucketName)
-	}
-	location, err := getStowLocation(ctx, cfg, p.secret)
-	if err != nil {
-		return nil, err
-	}
-	c, err := location.Container(bucketName)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to get bucket %s", bucketName)
-	}
-	dir := &directory{
-		path: "/",
-	}
-	hostEndPoint := p.hostEndPoint
-	if hostEndPoint == "" {
-		hostEndPoint = awsS3Endpoint(cfg.Region)
-	}
-	bucket := &bucket{
-		directory:    dir,
-		container:    c,
-		location:     location,
-		hostEndPoint: path.Join(hostEndPoint, c.ID()),
-		region:       cfg.Region,
+// GetRegionForBucket returns the region for a particular bucket. It does not
+// use the region set in the provider config is used as a hint, but may differ
```

Comment:
[minor] This sentence doesn't seem right
```suggestion
// GetRegionForBucket returns the region for a particular bucket.
// The region set in the provider config is used as a hint, but may differ
```

Reply:
sorry, I missed this comment. I'll address in a followup